codex/stores/maintenance.nim

```diff
@@ -10,92 +10,323 @@
+          checkpoint: currDatasetMd.checkpoint
+        )
+        return datasetMd.some
+      else:
+        raise newException(CatchableError, "DatasetMetadata for treeCid " & $treeCid & " not found")
+  )
+
+
+proc recordCheckpoint*(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  # Saves progress or deletes dataset metadata if progress > leavesCount
+  #
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
 
-proc runBlockCheck(self: BlockMaintainer): Future[void] {.async.} =
-  let expirations = await self.repoStore.getBlockExpirations(
-    maxNumber = self.numberOfBlocksPerInterval,
-    offset = self.offset
+  await self.metaDs.modify(key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        if currDatasetMd.expiry != datasetMd.expiry or currDatasetMd.manifestsCids != datasetMd.manifestsCids:
+          raise newException(CatchableError, "Change in expiry detected, interrupting maintenance for dataset with treeCid " & $treeCid)
+
+        if currDatasetMd.checkpoint.progress > datasetMd.checkpoint.progress:
+          raise newException(CatchableError, "Progress should be increasing only, treeCid " & $treeCid)
+
+        if currDatasetMd.leavesCount <= datasetMd.checkpoint.progress:
+          DatasetMetadata.none
+        else:
+          datasetMd.some
+      else:
+        raise newException(CatchableError, "Metadata for dataset with treeCid " & $treeCid & " not found")
   )
 
-  without iter =? expirations, err:
-    trace "Unable to obtain blockExpirations iterator from repoStore"
-    return
+proc deleteBatch(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  var datasetMd = datasetMd
+
+  datasetMd.checkpoint.timestamp = self.clock.now
+
+  if err =? (await self.recordCheckpoint(treeCid, datasetMd)).errorOption:
+    return failure(err)
+
+  logScope:
+    treeCid = treeCid
+    manifestsCids = datasetMd.manifestsCids
+    startingIndex = datasetMd.checkpoint.progress
+
+  trace "Deleting a batch of blocks", size = self.batchSize
+
+  var index = datasetMd.checkpoint.progress
+  while (index < datasetMd.checkpoint.progress + self.batchSize) and
+          (index < datasetMd.leavesCount):
+    if err =? (await self.blockStore.delBlock(treeCid, index)).errorOption:
+      error "Error deleting a block", msg = err.msg, index
+
+    index.inc
+
+    await sleepAsync(50.millis) # cooperative scheduling
+
+  if index >= datasetMd.leavesCount:
+    trace "All blocks deleted from a dataset", leavesCount = datasetMd.leavesCount
+
+    for manifestCid in datasetMd.manifestsCids:
+      if err =? (await self.blockStore.delBlock(manifestCid)).errorOption:
+        error "Error deleting manifest", cid = manifestCid
+
+    if err =? (await self.blockStore.delBlock(treeCid)).errorOption:
```

Comment:
What are we storing as blocks under `treeCid`?

Reply:
Noting, it's for future proofing because we had talks about storing trees directly as a blob (to use it for proofs). If you want I can remove it.

Reply:
Removing it.